File: packages/dotnet/test/Codec.Net.Tests/SafetyPolicyTests.cs
// SPDX-License-Identifier: MIT
using System.Net;
using System.Net.Http;
using System.Security.Cryptography;
using System.Text;
using System.Text.Json;
using Xunit;
using Codec;

namespace Codec.Net.Tests;

public class SafetyPolicyTests
{
    private const string Origin = "https://acme.test";

    private static SafetyPolicyDescriptor BuildValid() => new()
    {
        Id = "acme/strict-v3",
        Version = "1",
        Tokenizers = new[] { "meta-llama/llama-3" },
        Categories = new[]
        {
            new SafetyCategory { Name = "secrets", Action = "stop" },
            new SafetyCategory { Name = "pii", Action = "redact", Description = "Email and phone." },
        },
        Classifier = new SafetyClassifierBlock
        {
            Family = "llama-guard-3-1b",
            Host = "server",
            RequiresEngineFeatures = new[] { "logits_processor", "sampling_chain" },
        },
        RulesSummary = new SafetyRulesSummary
        {
            BannedTokenIdCount = 4128,
            RegexPatternCount = 47,
        },
        ClientHooks = new SafetyClientHooksBlock
        {
            PrefilterCategories = new[] { "secrets", "pii" },
            ClientClassifierFamily = "prompt-guard-86m",
        },
        PublishedAt = "2026-05-09T00:00:00Z",
    };

    private static byte[] Serialize(SafetyPolicyDescriptor d)
        => SafetyPolicy.CanonicalBytes(d);

    private static byte[] SerializeJson(object o)
        => JsonSerializer.SerializeToUtf8Bytes(o);

    // ── Validation ─────────────────────────────────────────────────────────

    [Fact]
    public void Validate_AcceptsMinimalValidDescriptor()
    {
        var bytes = Serialize(BuildValid());
        using var doc = JsonDocument.Parse(bytes);
        SafetyPolicy.Validate(doc.RootElement);  // no throw
    }

    [Fact]
    public void Validate_RejectsMissingRequiredFields()
    {
        using var doc = JsonDocument.Parse("{}");
        Assert.Throws<SafetyPolicyValidationException>(() => SafetyPolicy.Validate(doc.RootElement));
    }

    [Fact]
    public void Validate_RejectsBadCategoryName()
    {
        using var doc = JsonDocument.Parse(SerializeJson(new
        {
            id = "x/y",
            version = "1",
            tokenizers = new[] { "t" },
            categories = new[] { new { name = "BadCaps", action = "stop" } },
            classifier = new { family = "f" },
        }));
        Assert.Throws<SafetyPolicyValidationException>(() => SafetyPolicy.Validate(doc.RootElement));
    }

    [Fact]
    public void Validate_RejectsUnknownAction()
    {
        using var doc = JsonDocument.Parse(SerializeJson(new
        {
            id = "x/y",
            version = "1",
            tokenizers = new[] { "t" },
            categories = new[] { new { name = "secrets", action = "banhammer" } },
            classifier = new { family = "f" },
        }));
        Assert.Throws<SafetyPolicyValidationException>(() => SafetyPolicy.Validate(doc.RootElement));
    }

    [Fact]
    public void Validate_RejectsUnknownEngineFeature()
    {
        using var doc = JsonDocument.Parse(SerializeJson(new
        {
            id = "x/y",
            version = "1",
            tokenizers = new[] { "t" },
            categories = new[] { new { name = "secrets", action = "stop" } },
            classifier = new { family = "f", requires_engine_features = new[] { "weather_api" } },
        }));
        Assert.Throws<SafetyPolicyValidationException>(() => SafetyPolicy.Validate(doc.RootElement));
    }

    // ── Hashing ────────────────────────────────────────────────────────────

    [Fact]
    public void Hash_IsDeterministicForIdenticalInput()
    {
        var d = BuildValid();
        Assert.Equal(SafetyPolicy.Hash(d), SafetyPolicy.Hash(d));
        Assert.StartsWith("sha256:", SafetyPolicy.Hash(d));
        Assert.Equal(64, SafetyPolicy.Hash(d).Substring(7).Length);
    }

    [Fact]
    public void Hash_DiffersWhenCategoryActionChanges()
    {
        var d1 = BuildValid();
        var d2 = d1 with
        {
            Categories = new[]
            {
                new SafetyCategory { Name = "secrets", Action = "flag" },
                new SafetyCategory { Name = "pii", Action = "redact", Description = "Email and phone." },
            },
        };
        Assert.NotEqual(SafetyPolicy.Hash(d1), SafetyPolicy.Hash(d2));
    }

    [Fact]
    public void CanonicalBytes_2SpaceIndentTrailingNewline()
    {
        var raw = SafetyPolicy.CanonicalBytes(BuildValid());
        var text = Encoding.UTF8.GetString(raw);
        Assert.EndsWith("\n", text);
        Assert.Contains("\n  ", text);
        // Round-trips through JSON.
        using var doc = JsonDocument.Parse(text);
        Assert.Equal(JsonValueKind.Object, doc.RootElement.ValueKind);
    }

    // ── URL builders ───────────────────────────────────────────────────────

    [Fact]
    public void WellKnownPolicyUrl_PreservesSlashes()
    {
        Assert.Equal(
            "https://acme.example/.well-known/codec/policies/acme/strict-v3.json",
            SafetyPolicy.WellKnownPolicyUrl("https://acme.example/", "acme/strict-v3"));
    }

    [Fact]
    public void WellKnownPolicyUrl_RejectsTraversal()
    {
        Assert.Throws<SafetyPolicyDiscoveryException>(
            () => SafetyPolicy.WellKnownPolicyUrl("https://acme.example", "../etc"));
        Assert.Throws<SafetyPolicyDiscoveryException>(
            () => SafetyPolicy.WellKnownPolicyUrl("https://acme.example", "/abs"));
        Assert.Throws<SafetyPolicyDiscoveryException>(
            () => SafetyPolicy.WellKnownPolicyUrl("https://acme.example", "trailing/"));
    }

    [Fact]
    public void WellKnownPolicyUrl_RejectsBadCharset()
    {
        Assert.Throws<SafetyPolicyDiscoveryException>(
            () => SafetyPolicy.WellKnownPolicyUrl("https://acme.example", "Acme/Strict"));
    }

    [Fact]
    public void WellKnownPolicyHashUrl_UsesSha256Path()
    {
        var hex = new string('a', 64);
        Assert.Equal(
            $"https://acme.example/.well-known/codec/policies/sha256/{hex}.json",
            SafetyPolicy.WellKnownPolicyHashUrl("https://acme.example", hex));
    }

    [Fact]
    public void WellKnownPolicyHashUrl_RejectsMalformedHex()
    {
        Assert.Throws<SafetyPolicyDiscoveryException>(
            () => SafetyPolicy.WellKnownPolicyHashUrl("https://acme.example", "not-hex"));
    }

    // ── Loader (mock HttpClient) ───────────────────────────────────────────

    private static HttpClient MakeMockHttp(Dictionary<string, (HttpStatusCode status, byte[] body)> routes)
    {
        var handler = new MockHandler(routes);
        return new HttpClient(handler);
    }

    private sealed class MockHandler : HttpMessageHandler
    {
        private readonly Dictionary<string, (HttpStatusCode status, byte[] body)> _routes;
        public MockHandler(Dictionary<string, (HttpStatusCode, byte[])> routes) { _routes = routes; }
        protected override Task<HttpResponseMessage> SendAsync(
            HttpRequestMessage request, CancellationToken cancellationToken)
        {
            var url = request.RequestUri!.ToString();
            if (_routes.TryGetValue(url, out var route))
            {
                return Task.FromResult(new HttpResponseMessage(route.status)
                {
                    Content = new ByteArrayContent(route.body),
                });
            }
            return Task.FromResult(new HttpResponseMessage(HttpStatusCode.NotFound));
        }
    }

    [Fact]
    public async Task LoadAsync_FetchesAndValidates()
    {
        var url = $"{Origin}/policies/acme.json";
        var bytes = SafetyPolicy.CanonicalBytes(BuildValid());
        using var http = MakeMockHttp(new() { [url] = (HttpStatusCode.OK, bytes) });
        var d = await SafetyPolicy.LoadAsync(url, http: http);
        Assert.Equal("acme/strict-v3", d.Id);
    }

    [Fact]
    public async Task LoadAsync_VerifiesHash()
    {
        var url = $"{Origin}/policies/acme.json";
        var bytes = SafetyPolicy.CanonicalBytes(BuildValid());
        var hex = Convert.ToHexString(SHA256.HashData(bytes)).ToLowerInvariant();
        var goodHash = $"sha256:{hex}";
        using var http = MakeMockHttp(new() { [url] = (HttpStatusCode.OK, bytes) });
        var d = await SafetyPolicy.LoadAsync(url, hash: goodHash, http: http);
        Assert.Equal("acme/strict-v3", d.Id);

        var wrongHash = "sha256:" + new string('b', 64);
        using var http2 = MakeMockHttp(new() { [url] = (HttpStatusCode.OK, bytes) });
        await Assert.ThrowsAsync<SafetyPolicyHashMismatchException>(
            () => SafetyPolicy.LoadAsync(url, hash: wrongHash, http: http2));
    }

    // ── Discovery ──────────────────────────────────────────────────────────

    [Fact]
    public async Task DiscoverAsync_ResolvesInlineDescriptor()
    {
        var url = SafetyPolicy.WellKnownPolicyUrl(Origin, "acme/strict-v3");
        var bytes = SafetyPolicy.CanonicalBytes(BuildValid());
        using var http = MakeMockHttp(new() { [url] = (HttpStatusCode.OK, bytes) });
        var d = await SafetyPolicy.DiscoverAsync(Origin, "acme/strict-v3", http: http);
        Assert.Equal("acme/strict-v3", d.Id);
    }

    [Fact]
    public async Task DiscoverAsync_404RaisesNotFound()
    {
        using var http = MakeMockHttp(new());
        await Assert.ThrowsAsync<SafetyPolicyDiscoveryNotFoundException>(
            () => SafetyPolicy.DiscoverAsync(Origin, "acme/strict-v3", http: http));
    }

    [Fact]
    public async Task DiscoverAsync_WithHashHitsContentAddressedSibling()
    {
        var bytes = SafetyPolicy.CanonicalBytes(BuildValid());
        var hex = Convert.ToHexString(SHA256.HashData(bytes)).ToLowerInvariant();
        var hashUrl = SafetyPolicy.WellKnownPolicyHashUrl(Origin, hex);
        using var http = MakeMockHttp(new() { [hashUrl] = (HttpStatusCode.OK, bytes) });
        var d = await SafetyPolicy.DiscoverAsync(Origin, "acme/strict-v3", $"sha256:{hex}", http: http);
        Assert.Equal("acme/strict-v3", d.Id);
    }

    [Fact]
    public async Task DiscoverAsync_WithHashRejectsBytesMismatch()
    {
        var wrongHex = new string('c', 64);
        var hashUrl = SafetyPolicy.WellKnownPolicyHashUrl(Origin, wrongHex);
        // The fetched bytes hash to something OTHER than wrongHex.
        var bytes = SafetyPolicy.CanonicalBytes(BuildValid());
        using var http = MakeMockHttp(new() { [hashUrl] = (HttpStatusCode.OK, bytes) });
        await Assert.ThrowsAsync<SafetyPolicyHashMismatchException>(
            () => SafetyPolicy.DiscoverAsync(Origin, "acme/strict-v3", $"sha256:{wrongHex}", http: http));
    }

    [Fact]
    public async Task DiscoverAsync_RejectsInlineIdMismatch()
    {
        var url = SafetyPolicy.WellKnownPolicyUrl(Origin, "acme/strict-v3");
        var d = BuildValid() with { Id = "someone-else/v1" };
        var bytes = SafetyPolicy.CanonicalBytes(d);
        using var http = MakeMockHttp(new() { [url] = (HttpStatusCode.OK, bytes) });
        await Assert.ThrowsAsync<SafetyPolicyDiscoveryException>(
            () => SafetyPolicy.DiscoverAsync(Origin, "acme/strict-v3", http: http));
    }

    [Fact]
    public void RoundTrip_DescriptorBytesParse()
    {
        var d = BuildValid();
        var bytes = SafetyPolicy.CanonicalBytes(d);
        var d2 = SafetyPolicy.FromJson(bytes);
        Assert.Equal(d.Id, d2.Id);
        Assert.Equal(d.Categories.Count, d2.Categories.Count);
        Assert.Equal(d.Classifier.Family, d2.Classifier.Family);
    }
}

File: packages/java/src/test/java/ai/codec/SafetyPolicyTests.java
// SPDX-License-Identifier: MIT
package ai.codec;

import com.fasterxml.jackson.databind.ObjectMapper;
import org.junit.jupiter.api.Test;

import java.security.MessageDigest;
import java.util.HexFormat;
import java.util.List;

import static org.junit.jupiter.api.Assertions.*;

/**
 * Java parity tests for slice 11. Mirrors the TS / Python / Rust /
 * .NET suites — same shape, same assertions.
 */
class SafetyPolicyTests {

    private static final ObjectMapper M = new ObjectMapper();

    private static SafetyPolicyDescriptor buildValid() {
        SafetyPolicyDescriptor d = new SafetyPolicyDescriptor();
        d.id = "acme/strict-v3";
        d.version = "1";
        d.tokenizers = List.of("meta-llama/llama-3");

        SafetyPolicyDescriptor.Category c1 = new SafetyPolicyDescriptor.Category();
        c1.name = "secrets"; c1.action = "stop";
        SafetyPolicyDescriptor.Category c2 = new SafetyPolicyDescriptor.Category();
        c2.name = "pii"; c2.action = "redact"; c2.description = "Email and phone.";
        d.categories = List.of(c1, c2);

        SafetyPolicyDescriptor.ClassifierBlock cls = new SafetyPolicyDescriptor.ClassifierBlock();
        cls.family = "llama-guard-3-1b";
        cls.host = "server";
        cls.requiresEngineFeatures = List.of("logits_processor", "sampling_chain");
        d.classifier = cls;

        SafetyPolicyDescriptor.RulesSummary rs = new SafetyPolicyDescriptor.RulesSummary();
        rs.bannedTokenIdCount = 4128L;
        rs.regexPatternCount = 47L;
        d.rulesSummary = rs;

        SafetyPolicyDescriptor.ClientHooksBlock ch = new SafetyPolicyDescriptor.ClientHooksBlock();
        ch.prefilterCategories = List.of("secrets", "pii");
        ch.clientClassifierFamily = "prompt-guard-86m";
        d.clientHooks = ch;

        d.publishedAt = "2026-05-09T00:00:00Z";
        return d;
    }

    // ── Validation ─────────────────────────────────────────────────────────

    @Test
    void validate_acceptsMinimalValidDescriptor() throws Exception {
        var bytes = SafetyPolicy.canonicalBytes(buildValid());
        SafetyPolicy.validate(M.readTree(bytes));
    }

    @Test
    void validate_rejectsMissingRequiredFields() throws Exception {
        assertThrows(SafetyPolicyValidationException.class,
                () -> SafetyPolicy.validate(M.readTree("{}")));
    }

    @Test
    void validate_rejectsBadCategoryName() throws Exception {
        var bad = """
            {"id": "x/y", "version": "1", "tokenizers": ["t"],
             "categories": [{"name": "BadCaps", "action": "stop"}],
             "classifier": {"family": "f"}}
            """;
        assertThrows(SafetyPolicyValidationException.class,
                () -> SafetyPolicy.validate(M.readTree(bad)));
    }

    @Test
    void validate_rejectsUnknownAction() throws Exception {
        var bad = """
            {"id": "x/y", "version": "1", "tokenizers": ["t"],
             "categories": [{"name": "secrets", "action": "banhammer"}],
             "classifier": {"family": "f"}}
            """;
        assertThrows(SafetyPolicyValidationException.class,
                () -> SafetyPolicy.validate(M.readTree(bad)));
    }

    @Test
    void validate_rejectsUnknownEngineFeature() throws Exception {
        var bad = """
            {"id": "x/y", "version": "1", "tokenizers": ["t"],
             "categories": [{"name": "secrets", "action": "stop"}],
             "classifier": {"family": "f",
                            "requires_engine_features": ["weather_api"]}}
            """;
        assertThrows(SafetyPolicyValidationException.class,
                () -> SafetyPolicy.validate(M.readTree(bad)));
    }

    // ── Hashing ────────────────────────────────────────────────────────────

    @Test
    void hash_isDeterministicForIdenticalInput() {
        var d = buildValid();
        var a = SafetyPolicy.hash(d);
        var b = SafetyPolicy.hash(d);
        assertEquals(a, b);
        assertTrue(a.startsWith("sha256:"));
        assertEquals(64, a.substring(7).length());
    }

    @Test
    void hash_differsWhenCategoryActionChanges() {
        var d1 = buildValid();
        var d2 = buildValid();
        d2.categories.get(0).action = "flag";
        assertNotEquals(SafetyPolicy.hash(d1), SafetyPolicy.hash(d2));
    }

    @Test
    void canonicalBytes_2SpaceIndentTrailingNewline() {
        var raw = SafetyPolicy.canonicalBytes(buildValid());
        var text = new String(raw);
        assertTrue(text.endsWith("\n"));
        assertTrue(text.contains("\n  "));
        // round-trips through Jackson
        assertDoesNotThrow(() -> M.readTree(text));
    }

    // ── URL builders ───────────────────────────────────────────────────────

    @Test
    void wellKnownPolicyUrl_preservesSlashes() {
        assertEquals(
                "https://acme.example/.well-known/codec/policies/acme/strict-v3.json",
                SafetyPolicy.wellKnownPolicyUrl("https://acme.example/", "acme/strict-v3"));
    }

    @Test
    void wellKnownPolicyUrl_rejectsTraversal() {
        assertThrows(SafetyPolicyDiscoveryException.class,
                () -> SafetyPolicy.wellKnownPolicyUrl("https://acme.example", "../etc"));
        assertThrows(SafetyPolicyDiscoveryException.class,
                () -> SafetyPolicy.wellKnownPolicyUrl("https://acme.example", "/abs"));
        assertThrows(SafetyPolicyDiscoveryException.class,
                () -> SafetyPolicy.wellKnownPolicyUrl("https://acme.example", "trailing/"));
    }

    @Test
    void wellKnownPolicyUrl_rejectsBadCharset() {
        assertThrows(SafetyPolicyDiscoveryException.class,
                () -> SafetyPolicy.wellKnownPolicyUrl("https://acme.example", "Acme/Strict"));
    }

    @Test
    void wellKnownPolicyHashUrl_usesSha256Path() {
        var hex = "a".repeat(64);
        assertEquals(
                "https://acme.example/.well-known/codec/policies/sha256/" + hex + ".json",
                SafetyPolicy.wellKnownPolicyHashUrl("https://acme.example", hex));
    }

    @Test
    void wellKnownPolicyHashUrl_rejectsMalformedHex() {
        assertThrows(SafetyPolicyDiscoveryException.class,
                () -> SafetyPolicy.wellKnownPolicyHashUrl("https://acme.example", "not-hex"));
    }

    // ── Round-trip ─────────────────────────────────────────────────────────

    @Test
    void roundTrip_descriptorBytesParse() {
        var d = buildValid();
        var bytes = SafetyPolicy.canonicalBytes(d);
        var d2 = SafetyPolicy.fromJson(bytes);
        assertEquals(d.id, d2.id);
        assertEquals(d.categories.size(), d2.categories.size());
        assertEquals(d.classifier.family, d2.classifier.family);
    }

    @Test
    void fromJson_rejectsBadShape() {
        assertThrows(SafetyPolicyValidationException.class,
                () -> SafetyPolicy.fromJson("{}".getBytes()));
    }

    // ── Cross-stack hash sanity ────────────────────────────────────────────

    @Test
    void hash_matchesDirectSha256OfCanonicalBytes() throws Exception {
        var d = buildValid();
        var bytes = SafetyPolicy.canonicalBytes(d);
        var direct = HexFormat.of().formatHex(MessageDigest.getInstance("SHA-256").digest(bytes));
        var via = SafetyPolicy.hash(d).substring("sha256:".length());
        assertEquals(direct, via);
    }
}
